codec-http/src/main/java/io/netty/handler/codec/http/ReadOnlyHttpHeaders.java

```diff
@@ -0,0 +1,322 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.handler.codec.CharSequenceValueConverter;
+import io.netty.util.AsciiString;
+import io.netty.util.internal.UnstableApi;
+
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import static io.netty.handler.codec.http.DefaultHttpHeaders.HttpNameValidator;
+
+@UnstableApi
+public final class ReadOnlyHttpHeaders extends HttpHeaders {
+    private final AsciiString[] keyValuePairs;
+
+    public ReadOnlyHttpHeaders(boolean validateHeaders, AsciiString... keyValuePairs) {
+        if ((keyValuePairs.length & 1) != 0) {
+            throw newInvalidArraySizeException();
+        }
+        if (validateHeaders) {
+            validateHeaders(keyValuePairs);
+        }
+        this.keyValuePairs = keyValuePairs;
```

Comment:
do we need to call `clone()` to ensure we make a copy as people could still hold an reference on the original one ?

Reply:
We do not do this for `ReadOnlyHttp2Headers`, and instead leave it up to the user if they want to pay the cost to avoid external modifications. Let me clarify in the java doc.